File: src/main/java/org/nyxlang/interpreter/visitor/ProgramVisitor.java
package org.nyxlang.interpreter.visitor;

import org.nyxlang.interpreter.IInterpreter;
import org.nyxlang.interpreter.exception.VisitorException;
import org.nyxlang.parser.ast.INode;
import org.nyxlang.parser.ast.ProgramNode;

/**
 * Interprets the program logic.
 */
public class ProgramVisitor implements IVisitor {

    @Override
    public boolean matches(INode node) {
        return ProgramNode.class == node.getClass();
    }

    @Override
    public Object visit(IInterpreter interpreter, INode node) throws VisitorException {
        var programNode = (ProgramNode) node;
        return interpreter.evalNode(programNode.getStatements());
    }
}

File: src/main/java/org/nyxlang/interpreter/visitor/ProgramVisitor.kt
package org.nyxlang.interpreter.visitor

import org.nyxlang.interpreter.IInterpreter
import org.nyxlang.parser.ast.INode
import org.nyxlang.parser.ast.ProgramNode

/**
 * Interprets the program logic.
 */
class ProgramVisitor : IVisitor {
    override fun matches(node: INode) = ProgramNode::class == node::class
    override fun visit(interpreter: IInterpreter, node: INode): Any? {
        val programNode = node as ProgramNode
        return interpreter.evalNode(programNode.statements)
    }
}